contracts/Pool.sol

```diff
@@ -435,36 +435,34 @@ library ReceiverWeightsImpl {
                     // Somehow it's ~1500 gas cheaper than `delete self[next]`
                     self.data[next].next = ADDR_UNINITIALIZED;
                     next = newNext;
-                    if (next == ADDR_END) break;
+                    if (next == ADDR_END) {
+                        // Removing the last item on the list, clear the storage
+                        if (current == ADDR_ROOT) next = ADDR_UNINITIALIZED;
+                        break;
+                    }
                     weight = self.data[next].weight;
                 } while (weight == 0);
                 // link the previous non-zero element with the next non-zero element
                 // or ADDR_END if it became the last element on the list
                 self.data[current].next = next;
             }
         }
+        if (next == ADDR_END) next = ADDR_ROOT;
```

Comment:
Is there a reason we’re distinguishing between `ADDR_END` and `ADDR_ROOT`? This pattern here is common with `nextWeight` and `setWeight`. It may be possible that we don’t need to distinguish these.

Reply:
Unfortunately we need both of them. They differentiate between a slot for a weight, which may be empty, but is attached to the linked list (`ADDR_END`) and a slot, which is empty and not attached (`ADDR_ROOT` aka `ADDR_UNINITIALIZED`). It prevents the list from creating a cycle when the last item is zeroed and then assigned a value.

Reply:
I think I’ve finally understood it. The distinction only becomes important in `setWeights` where we need to decide whether to insert or to update in place.